File: src/ws/registry.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{mpsc, RwLock};
use uuid::Uuid;

/// A handle to send messages to a connected device's WebSocket
pub type DeviceSender = mpsc::Sender<String>;

#[derive(Clone, Default)]
pub struct WsRegistry {
    inner: Arc<RwLock<HashMap<Uuid, DeviceSender>>>,
}

impl WsRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn insert(&self, device_id: Uuid, tx: DeviceSender) {
        self.inner.write().await.insert(device_id, tx);
    }

    pub async fn remove(&self, device_id: Uuid) {
        self.inner.write().await.remove(&device_id);
    }

    pub async fn is_connected(&self, device_id: Uuid) -> bool {
        self.inner.read().await.contains_key(&device_id)
    }

    pub async fn send(&self, device_id: Uuid, msg: String) -> bool {
        if let Some(tx) = self.inner.read().await.get(&device_id) {
            tx.send(msg).await.is_ok()
        } else {
            false
        }
    }
}

File: server/src/ws/registry.ts
import type { WebSocket } from 'ws'

export class WsRegistry {
  private connections = new Map<string, WebSocket>()

  insert(deviceId: string, ws: WebSocket) {
    this.connections.set(deviceId, ws)
  }

  remove(deviceId: string) {
    this.connections.delete(deviceId)
  }

  isConnected(deviceId: string): boolean {
    return this.connections.has(deviceId)
  }

  send(deviceId: string, msg: string): boolean {
    const ws = this.connections.get(deviceId)
    if (ws && ws.readyState === ws.OPEN) {
      ws.send(msg)
      return true
    }
    return false
  }

  connectedCount(): number {
    return this.connections.size
  }
}
